File: android/app/src/main/java/app/releaf/mobile/features/activity/ActivityScreen.kt
/*
 * ActivityScreen.kt
 *
 * Full-screen activity log — same data source as the home timeline
 * card, just a bigger window (200 rows vs 5) and grouped by day so
 * the user can scan back through what they've touched.
 *
 * Drill-in screen reached from the home tab's "See all" link. Header
 * mirrors the home tab's vibe: tappable LeafEyebrow back-link
 * ("🌱 HOME · ACTIVITY") + a serif editorial title in 32sp serif.
 */

package app.releaf.mobile.features.activity

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.viewmodel.compose.viewModel
import app.releaf.mobile.data.activity.ActivityItem
import app.releaf.mobile.features.home.accentFor
import app.releaf.mobile.features.home.labelFor
import app.releaf.mobile.features.home.paletteFor
import app.releaf.mobile.ui.components.ActivityEntry
import app.releaf.mobile.ui.components.ActivityProminence
import app.releaf.mobile.ui.components.ActivityTimeline
import app.releaf.mobile.ui.components.LeafEyebrow
import app.releaf.mobile.ui.components.relativeTimeAgo
import app.releaf.mobile.ui.theme.AppAccent
import app.releaf.mobile.ui.theme.AppColors
import app.releaf.mobile.ui.theme.AppRadius
import app.releaf.mobile.ui.theme.AppSpacing
import app.releaf.mobile.ui.theme.AppTypography
import app.releaf.mobile.ui.theme.TimelineStyle
import app.releaf.mobile.ui.theme.UiPreferences
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter

@Composable
fun ActivityScreen(
    onBack: () -> Unit,
    modifier: Modifier = Modifier,
    viewModel: RecentActivityViewModel = viewModel(
        factory = RecentActivityViewModel.factory(RecentActivityViewModel.FULL_LIMIT),
    ),
) {
    val items by viewModel.items.collectAsState()
    // Timeline style follows the same Settings → Timeline preference
    // the home tab's HomeTimelineCard reads. Classic = the simple
    // rounded-row list with a colored dot + relative time. Bramble =
    // the serpentine-vine + flower marker timeline.
    val context = LocalContext.current
    val prefs = remember(context) { UiPreferences.get(context) }
    val style by prefs.state.collectAsState()

    Column(
        modifier = modifier
            .fillMaxSize()
            .background(AppColors.Canvas),
    ) {
        TopBar(onBack = onBack, count = items.size)

        if (items.isEmpty()) {
            Box(
                Modifier
                    .fillMaxSize()
                    .padding(AppSpacing.s8),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    text = "Nothing here yet — your edits will show up as you make them.",
                    style = AppTypography.Body,
                    color = AppColors.TextSecondary,
                    textAlign = TextAlign.Center,
                )
            }
        } else {
            // Group rows by local date so the user gets a "Today /
            // Yesterday / Mon Apr 21" heading separating them.
            val grouped = items.groupBy { localDateOf(it.timestamp) }
            LazyColumn(
                modifier = Modifier.fillMaxSize(),
                contentPadding = PaddingValues(
                    start  = AppSpacing.s4,
                    end    = AppSpacing.s4,
                    top    = AppSpacing.s2,
                    bottom = AppSpacing.s10,
                ),
                verticalArrangement = Arrangement.spacedBy(AppSpacing.s4),
            ) {
                when (style.timelineStyle) {
                    TimelineStyle.Classic -> {
                        // Classic rendering: each group → eyebrow header
                        // + a stack of dot-led rows on individual cards.
                        grouped.forEach { (date, rows) ->
                            item(key = "header:${date}") {
                                SectionHeader(label = headingFor(date))
                            }
                            items(rows, key = { "classic:${it.id}" }) { row ->
                                ActivityRow(row)
                            }
                        }
                    }
                    TimelineStyle.Bramble -> {
                        // Bramble rendering: each group → its own
                        // serpentine-vine ActivityTimeline card with the
                        // date as the eyebrow header.
                        grouped.forEach { (date, rows) ->
                            item(key = "section:$date") {
                                ActivityTimeline(
                                    header  = headingFor(date).uppercase(),
                                    entries = rows.mapIndexed { index, item ->
                                        ActivityEntry(
                                            id         = item.id,
                                            date       = relativeTimeAgo(item.timestamp),
                                            title      = labelFor(item),
                                            preview    = item.context,
                                            theme      = paletteFor(item.kind),
                                            prominence = if (index == 0) ActivityProminence.Featured
                                                         else ActivityProminence.Routine,
                                        )
                                    },
                                )
                            }
                        }
                    }
                }
            }
        }
    }
}

@Composable
private fun SectionHeader(label: String) {
    Text(
        text     = label.uppercase(),
        style    = AppTypography.Eyebrow,
        color    = AppAccent.primary,
        modifier = Modifier.padding(top = AppSpacing.s2, bottom = AppSpacing.s1),
    )
}

@Composable
private fun ActivityRow(item: ActivityItem) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(AppRadius.md))
            .background(AppColors.CardSolid)
            .border(1.dp, AppColors.BorderDefault, RoundedCornerShape(AppRadius.md))
            .padding(AppSpacing.s4),
        horizontalArrangement = Arrangement.spacedBy(AppSpacing.s3),
        verticalAlignment     = Alignment.Top,
    ) {
        Box(
            modifier = Modifier
                .padding(top = 4.dp)
                .size(10.dp)
                .clip(CircleShape)
                .background(accentFor(item.kind)),
        )
        Column(
            modifier            = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(2.dp),
        ) {
            Text(
                text     = labelFor(item),
                style    = AppTypography.Body,
                color    = AppColors.TextPrimary,
                maxLines = 2,
            )
            Text(
                text  = relativeTimeAgo(item.timestamp),
                style = AppTypography.Meta,
                color = AppColors.TextSecondary,
            )
        }
    }
}

@Composable
private fun TopBar(onBack: () -> Unit, count: Int) {
    // Eyebrow + serif title rhythm — same shape as the notebook detail
    // and chapter detail headers we recently aligned, and the home
    // tab's "RELEAF / greeting" header. Eyebrow links back; the count
    // chip sits on the trailing edge of the title row.
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(
                start  = AppSpacing.s4,
                end    = AppSpacing.s4,
                top    = AppSpacing.s4,
                bottom = AppSpacing.s3,
            ),
        verticalArrangement = Arrangement.spacedBy(AppSpacing.s3),
    ) {
        LeafEyebrow(
            label    = "home · activity",
            modifier = Modifier.clickable(onClick = onBack),
        )
        Row(
            verticalAlignment     = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(AppSpacing.s2),
            modifier              = Modifier.fillMaxWidth(),
        ) {
            Text(
                text     = "Recent edits",
                style    = TextStyle(
                    fontFamily = FontFamily.Serif,
                    fontSize   = 32.sp,
                ),
                color    = AppColors.TextPrimary,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
                modifier = Modifier.weight(1f),
            )
            if (count > 0) {
                Box(
                    modifier = Modifier
                        .clip(RoundedCornerShape(AppRadius.pill))
                        .background(AppAccent.soft)
                        .padding(horizontal = AppSpacing.s3, vertical = AppSpacing.s1),
                ) {
                    Text(
                        text  = "$count",
                        style = AppTypography.Tag,
                        color = AppAccent.deep,
                    )
                }
            }
        }
    }
}

/* ---------- date helpers ---------- */

private fun localDateOf(iso: String): LocalDate =
    runCatching {
        Instant.parse(iso).atZone(ZoneId.systemDefault()).toLocalDate()
    }.getOrElse { LocalDate.now() }

private fun headingFor(date: LocalDate): String {
    val today = LocalDate.now()
    return when (date) {
        today                  -> "Today"
        today.minusDays(1)     -> "Yesterday"
        else                   -> DateTimeFormatter.ofPattern("EEE MMM d").format(date)
    }
}

File: ios/Releaf/Features/Activity/ActivityScreen.swift
/*
 * ActivityScreen.swift
 *
 * Full-screen activity log — same data source as the Home timeline
 * card, just a wider window (full limit vs HOME_LIMIT) so the user
 * can scroll back through everything they've touched.
 *
 * Mirrors `features/activity/ActivityScreen.kt` on Android. Uses the
 * Bramble timeline visual treatment (vine + flowers) as the primary
 * surface since the user opted into it via Settings; the Classic
 * dot-on-rail rendering is still available on Home.
 */

import SwiftUI
import ReleafData
import ReleafDesignSystem

public struct ActivityScreen: View {
    @StateObject private var viewModel: RecentActivityViewModel
    @ObservedObject private var prefs: UiPreferences

    public init(userId: String, prefs: UiPreferences = .shared) {
        _viewModel = StateObject(
            wrappedValue: RecentActivityViewModel(
                userId: userId,
                maxItems: 200
            )
        )
        self.prefs = prefs
    }

    public var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: AppSpacing.s4) {
                header
                if viewModel.items.isEmpty {
                    emptyState
                } else {
                    ActivityTimeline(
                        entries: timelineEntries,
                        header: "All activity"
                    )
                }
                Spacer(minLength: AppSpacing.s10)
            }
            .padding(.horizontal, AppSpacing.s5)
            .padding(.top, AppSpacing.s5)
        }
        .background(AppColors.canvas.ignoresSafeArea())
        .onAppear { viewModel.start() }
        .onDisappear { viewModel.stop() }
    }

    // MARK: - Pieces

    private var header: some View {
        VStack(alignment: .leading, spacing: AppSpacing.s1) {
            Text("ACTIVITY")
                .font(AppText.eyebrow)
                .tracking(AppLetterSpacing.eyebrow)
                .foregroundStyle(AppColors.themeGreenDeep)
            Text("Timeline")
                .font(AppText.editorialTitle)
                .foregroundStyle(AppColors.textPrimary)
            Text("\(viewModel.items.count) item\(viewModel.items.count == 1 ? "" : "s")")
                .font(AppText.meta)
                .foregroundStyle(AppColors.textSecondary)
        }
    }

    private var emptyState: some View {
        Text("No activity yet — start by adding a note.")
            .font(AppText.body)
            .foregroundStyle(AppColors.textSecondary)
            .padding(.vertical, AppSpacing.s6)
            .frame(maxWidth: .infinity)
    }

    private var timelineEntries: [ActivityEntry] {
        viewModel.items.enumerated().map { index, item in
            ActivityEntry(
                id: UUID(uuidString: item.id) ?? UUID(),
                date: relativeTimeAgo(item.timestamp),
                title: labelFor(item),
                preview: item.context,
                theme: paletteFor(item.kind),
                prominence: index == 0 ? .featured : .routine
            )
        }
    }
}

// MARK: - Mappings

/// Mirrors the Home-card mapping. Sub-event kinds (photo, scan, etc.)
/// route to the same palette as on the timeline card.
private func paletteFor(_ kind: ActivityKind) -> AccentPaletteID {
    switch kind {
    case .notepadEntry, .voice:               return .coral
    case .notebook, .page, .chapter:          return .green
    case .photo, .scan, .todo:                return .yellow
    case .contact, .location:                 return .dry
    }
}

/// "<verb> <thing> · <title>", e.g. "Updated note · Morning notes".
private func labelFor(_ item: ActivityItem) -> String {
    let verb: String = {
        switch item.action {
        case .created:  return "Created"
        case .updated:  return "Updated"
        case .deleted:  return "Deleted"
        case .restored: return "Restored"
        case .merged:   return "Merged"
        case .moved:    return "Moved"
        }
    }()
    let noun: String = {
        switch item.kind {
        case .notepadEntry: return "note"
        case .page:         return "page"
        case .chapter:      return "chapter"
        case .notebook:     return "notebook"
        case .photo:        return "photo"
        case .scan:         return "scan"
        case .voice:        return "voice note"
        case .todo:         return "todo"
        case .contact:      return "contact"
        case .location:     return "location"
        }
    }()
    return "\(verb) \(noun) · \(item.title)"
}

private func relativeTimeAgo(_ iso: String) -> String {
    let f = ISO8601DateFormatter()
    f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    let parsed = f.date(from: iso) ?? {
        f.formatOptions = [.withInternetDateTime]
        return f.date(from: iso)
    }()
    guard let date = parsed else { return iso }
    let delta = max(0, Int(Date().timeIntervalSince(date)))
    if delta < 60         { return "just now" }
    if delta < 3600       { return "\(delta / 60)m ago" }
    if delta < 86_400     { return "\(delta / 3600)h ago" }
    if delta < 2 * 86_400 { return "yesterday" }
    if delta < 7 * 86_400 { return "\(delta / 86_400) days ago" }
    let fmt = DateFormatter()
    fmt.dateFormat = "MMM d"
    return fmt.string(from: date)
}
